### PR TITLE
[SPARK-32332][SQL] Make it possible to implement columnar exchanges

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This PR adds abstract classes for shuffle and broadcast, so that users can provide their columnar implementations.

This PR updates several places to use the abstract exchange classes, and also update `AdaptiveSparkPlanExec` so that the columnar rules can see exchange nodes.

This is an alternative of https://github.com/apache/spark/pull/29134 .
Close https://github.com/apache/spark/pull/29134

### Why are the changes needed?

To allow columnar exchanges.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @andygrove  @tgravescs @maryannxue

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 1]
```diff
@@ -376,10 +382,14 @@ case class AdaptiveSparkPlanExec(
   private def newQueryStage(e: Exchange): QueryStageExec = {
     val optimizedPlan = applyPhysicalRules(e.child, queryStageOptimizerRules)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:383]: it's important to hide the exchange nodes from the stage optimizer rules, as that's the assumption of these rules.

Then we don't need https://github.com/apache/spark/pull/29134/files#diff-a30c7a6fcdcdd13e57135fd04d05f3b7R115

**Comment by andygrove** [comment]
Thanks @cloud-fan I will test our AQE POC with these changes.

**Comment by andygrove** [comment]
Thanks @cloud-fan. I have tested these changes both with Spark 3.1 and also back portred to the 3.0 branch and everything is working well, so LGTM.

I wish I had thought to separate the rules out into stage creation and post-stage creation. That made things much simpler.

cc @tgravescs

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 2]
```diff
@@ -376,10 +382,14 @@ case class AdaptiveSparkPlanExec(
...
     val queryStage = e match {
-      case s: ShuffleExchangeExec =>
-        ShuffleQueryStageExec(currentStageId, s.copy(child = optimizedPlan))
+      case s: ShuffleExchangeLike =>
```
Comment (tgravescs) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:385]: this should be able to stay ShuffleExchangeExec or be consistent and change braodcast below to be BroadcastExchangeLike

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeLocalShuffleReader.scala [hunk 3]
```diff
@@ -78,10 +79,11 @@ case class OptimizeLocalShuffleReader(conf: SQLConf) extends Rule[SparkPlan] {
...
       shuffleStage: ShuffleQueryStageExec,
       advisoryParallelism: Option[Int]): Seq[ShufflePartitionSpec] = {
-    val shuffleDep = shuffleStage.shuffle.shuffleDependency
-    val numReducers = shuffleDep.partitioner.numPartitions
+    val numReducers = shuffleStage.shuffle.numPartitions
     val expectedParallelism = advisoryParallelism.getOrElse(numReducers)
-    val numMappers = shuffleDep.rdd.getNumPartitions
+    val mapOutputTracker = SparkEnv.get.mapOutputTracker.asInstanceOf[MapOutputTrackerMaster]
+    val numMappers =
+      mapOutputTracker.shuffleStatuses(shuffleStage.mapStats.get.shuffleId).mapStatuses.length
```
Comment (tgravescs) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeLocalShuffleReader.scala:86]: personally I'm not fond of us reaching into mapOutputTracker and looking at its internal datastructures. This I believe was exposed for testing purposes not for people to use it directly.  I would rather see either a utility function made or we do logic like we were before.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala [hunk 4]
```diff
@@ -40,13 +41,49 @@ import org.apache.spark.sql.types.StructType
...
 
+/**
+ * Common trait for all shuffle exchange implementations to facilitate pattern matching.
+ */
+trait ShuffleExchangeLike extends Exchange {
+
+  /**
+   * Returns the number of mappers of this shuffle.
+   */
+  def numMappers: Int
```
Comment (tgravescs) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala:52]: need to update the tests fto have this

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 5]
```diff
@@ -376,10 +382,14 @@ case class AdaptiveSparkPlanExec(
...
     val optimizedPlan = applyPhysicalRules(e.child, queryStageOptimizerRules)
     val queryStage = e match {
-      case s: ShuffleExchangeExec =>
-        ShuffleQueryStageExec(currentStageId, s.copy(child = optimizedPlan))
-      case b: BroadcastExchangeExec =>
-        BroadcastQueryStageExec(currentStageId, b.copy(child = optimizedPlan))
+      case s: ShuffleExchangeLike =>
+        val newShuffle = applyPhysicalRules(
+          s.withNewChildren(Seq(optimizedPlan)), postStageCreationRules)
+        ShuffleQueryStageExec(currentStageId, newShuffle)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:388]: Do we need to make sure `newShuffle` is still `ShuffleExchangeLike`? Seems custom rules possibly break it?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala [hunk 6]
```diff
@@ -40,13 +41,49 @@ import org.apache.spark.sql.types.StructType
...
+   */
+  def numPartitions: Int
+
+  /**
+   * Returns whether the shuffle partition number can be changed.
+   */
+  def canChangeNumPartitions: Boolean
+
+  /**
+   * The asynchronous job that prepares the materializes the shuffle.
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala:65]: This reads weird. A broken sentence?

**Comment by tgravescs** [comment]
test this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:392]: added a check

**Review by tgravescs — APPROVED** [review]
looks good, pending jenkins

**Review by viirya — APPROVED** [review]

**Comment by tgravescs** [comment]
test this please

**Comment by tgravescs** [comment]
I merged this to master, unfortunately wouldn't pick clean to branch-3.0.

@cloud-fan would you want to put up PR for branch-3.0? Otherwise Andy or myself can.

**Comment by cloud-fan** [comment]
Does it qualify a backport? It's kind of a new feature.

**Comment by tgravescs** [comment]
I went back and forth on that as I can see it both ways.  I ended up filing it as a bug as it prevents us from properly doing columnar processing (introduced in 3.0.0) with AQE.  I also thought it was all internal and a fairly isolated path and was hoping to get it into 3.0.1  if others didn't disagree since users are starting to to use AQE in 3.0. what do you think?

**Comment by cloud-fan** [comment]
I agree it's not that risky(the code is the same as before, just with one more abstraction layer: exchange-like). I don't have a strong opinion here. @andygrove can you help to create a backport PR?

**Comment by andygrove** [comment]
Sure, I'll create a PR this morning. Thanks @cloud-fan and @tgravescs.

**Pull request closed without merging**